### PR TITLE
[Patch 3/4] Network rework: Packet fixes & cleanups

### PR DESCRIPTION
**Pull request by nerzhul**
Cleanups:
* remove unhandled handlers and use deprecated handler
* remove some useless functions

Packet fixes:
* Rewrite TOCLIENT_DENYACCESS packet properly
* Fix DELETE_PARTICLESPAWNER
* Rewrite TOSERVER_INIT packet to use a dynamic size
* Rewrite TOSERVER_PASSWORD to use dynamic size

Misc:
* Connection::Receive function now uses NetworkPacket.

### CODE REVIEW

**Comment by nerzhul** [comment]
This PR is a guide to the protocol rework. discussions can be engaged to improve the network for everybody.

**Comment by Zeno-** [comment]
Is this backwards compatible?

**Comment by nerzhul** [comment]
No, this is impossible. As i said patch 3/4 break the protocole, we don't have choice. _Init packet must be reworked.

**Comment by Zeno-** [comment]
Can you tag it as such and make that clear?

**Comment by nerzhul** [comment]
I also think with the next release (> 0.4 branch) we must rework all the networking layer and the networking protocol for a better network.

I added the tag, i missed it :)

**Comment by est31** [comment]
#2334 contains changes to make server list aware of the protocol change. Either we release 0.4.12 (with my PR merged into it) before this gets merged, or we live with the master list broken in meantime.

**Comment by Zeno-** [comment]
This won't be in 0.4.12 (it will be it whatever version is after 0.4.12)

**Comment by est31** [comment]
sorry, second sentence shouldn't have gone into this issue. It was more targeted at #2334.

**Comment by ShadowNinja** [comment]
If we break the protocol, I think we should fix authentication at the same time.
@nerzhul, can you look into implementing SRP?

**Comment by Zeno-** [comment]
Related to @ShadowNinja's comment, I also mentioned to @nerzhul that "compression strategy" would be a useful addition to the protocol so that if a different compression strategy is added in the future it could be done seamlessly and without breaking protocol again.

**Comment by nerzhul** [comment]
@ShadowNinja agree with auth problems but after network rework, i think. Because here i'm focused on protocol problems, but auth is very important too.

**Comment by nerzhul** [comment]
@zeno- i'll add a compression byteflag in _init packet

**Comment by nerzhul** [comment]
@Zeno- we now have a Byteflag in _INIT packet which is set into client session (note: this clientiface session will be removed in 4/4, using a new an proper client session)

**Comment by est31** [comment]
I dunno about previous practices, but shouldn't this adjust `LATEST_PROTOCOL_VERSION`, along with `*_PROTOCOL_VERSION_MIN` adjustments? --> #2334 only provides a solution for bumped protocol versions.

**Comment by nerzhul** [comment]
Protocol version 25 declared. I also cleanup a unused define.

**Comment by ShadowNinja** [comment]
Compression flags aren't really right since it doesn't make sense to have `NONE` and `ZLIB` (and in the future eg, `GZIP`) both enabled at the same time.  Also, zlib isn't actually a compression method, it's a library that implements the `deflate` compression method.

Is a compression field really needed though?  Can't we just increment the serialization version or protocol version instead if we ever change that.

Also, I think player names are currently fixed-width in their class.  Be carefull you're not introducing a buffer overflow!  The right way to fix that is to make the size dynamic of course, but you'll have to add a (configurable) max playername size to prevent insanity like 65535-character player names (although it should be technically possible to have at least 255-character player names).

**Comment by nerzhul** [comment]
@ShadowNinja , playername size is already check with if (playerName.size() > PLAYERNAME_SIZE) {

I'll look at your other comments soon, but 0.4.12 release must be ready, and @celeron55 wants us to resolve all blocker issues

**Comment by ShadowNinja** [comment]
@nerzhul Yes, I see that, but there really shouldn't be a fixed size to begin with.  It invites buffer overflows (you already introduced one, although you later fixed it), it's inflexible (wastes space for short names, doesn't allow long ones), and it's harder to work with.

**Comment by nerzhul** [comment]
i'll look at all your comments today and improve everything :)

**Comment by Zeno-** [comment]
@ShadowNinja 
> Is a compression field really needed though? Can't we just increment the serialization version or protocol version instead if we ever change that.

Yes, it's needed so that new compression methods can be introduced without breaking the protocol.  It's not unreasonable to expect the server to support more than one compression strategy.

Also, currently serialized text stored in leveldb databases is zlib compressed; however leveldb already compresses using snappy, to it's compressed twice...

**Comment by est31** [comment]
While we're on it, can we remove [this deprecated functionality](https://github.com/nerzhul/minetest/blob/143b254b7bc1c8247e8204e6a8466c2a6cabe840/src/network/networkprotocol.h#L150) too?

**Comment by sfan5** [comment]
> however leveldb already compresses using snappy, so it's compressed twice...

`libsnappy` is optional when building LevelDB; e.g. all the Windows builds use a LevelDB without `libsnappy` (IIRC)

**Comment by Zeno-** [comment]
@sfan5 is that a reason not to support different compression methods (or even no compression)?

**Comment by sfan5** [comment]
This PR adds compression to network communications, that has nothing to do with MapBlock compression.

**Comment by nerzhul** [comment]
@sfan5 no, it add only a byteflag for server to know supported compression for client, which permit to change the compressions method in some packets which are using zlib, if we want. But at this time it's only informative.

File: src/client.cpp [hunk 1]
```diff
@@ -395,27 +395,18 @@ void Client::step(float dtime)
...
 
-			char pName[PLAYERNAME_SIZE];
-			char pPassword[PASSWORD_SIZE];
+			NetworkPacket* pkt = new NetworkPacket(TOSERVER_INIT, 0);
 
-			snprintf(pName, PLAYERNAME_SIZE, "%s", myplayer->getName());
-			snprintf(pPassword, PASSWORD_SIZE, "%s", m_password.c_str());
+			// Set the supported compression modes here (8 modes possible)
+			u8 compression_mode = 0;
+			compression_mode |= (1 << NETPROTO_COMPRESSION_NONE);
```
Comment (Zeno-) [inline src/client.cpp]: I don't think that no compression needs a bit; i.e. 0 would be no compression. Therefore until another compression method is supported, the default value of compression_mode should simply be `(1 << NETPROTO_COMPRESSION_ZLIB)`

File: src/clientiface.h [hunk 2]
```diff
@@ -308,11 +310,14 @@ class RemoteClient
...
 	}
 
+	void setSupportedCompressionModes(u8 byteFlag);
+
 	/* read version information */
 	u8 getMajor() { return m_version_major; }
 	u8 getMinor() { return m_version_minor; }
 	u8 getPatch() { return m_version_patch; }
 	std::string getVersion() { return m_full_version; }
+	bool supportedCompression[NETPROTO_COMPRESSION_MAX];
```
Comment (Zeno-) [inline src/clientiface.h]: Is this just for convenience? I.e. why can't it just be left as a bitfield?

Comment (nerzhul) [inline src/clientiface.h]: We can keep the byteflag, yes, not bad

**Comment by nerzhul** [comment]
Merge #2261

**Comment by nerzhul** [comment]
As said on IRC, merged in dev-0.5, please test.
Commit c106df68bceae0a8c5d8513c331e3bf035462fa2

**Pull request closed without merging**